File: MigrationVito/C/TestNIDAQmxAO_Finite.c
#define _USE_MATH_DEFINES

#include <NIDAQmx.h>
#include <math.h>
#include <stdio.h>
#include <time.h>

void testCode() {
    // DAQmxCreateAOVoltageChan options 
    const char* physicalChannel = "PXI1Slot5/ao0";
    const char* nameToAssignChannel = "AO Voltage";
    const float64 minVal = -10;
    const float64 maxVal = 10;
    const int32 units = DAQmx_Val_Volts;
    const char* customScaleName = "";

    // DAQmxCfgSampClkTiming options
    const char* source = "OnboardClock";
    const float64 rate = 3000000;
    const int32 activeEdge = DAQmx_Val_Rising;
    const int32 sampleMode = DAQmx_Val_FiniteSamps;
    const uInt64 sampsPerChanToAquire = 30000;

    // wave form configuration
    const double freq = 1000;
    const double amp = 5;

    // DAQmxWriteAnalogF64 options
    const bool32 autoStart = 0;
    const float64 timeout = 10;
    const bool32 dataLayout = DAQmx_Val_GroupByChannel;
    float64 writeArray[sampsPerChanToAquire];
    for (int i = 0; i < sampsPerChanToAquire; i++)
        writeArray[i] = amp * sin(2.0*M_PI * freq * (double)i / rate);
    int32 sampsPerChanWritten;

    // DAQmxWaitUntilTaskDone options
    const float64 timeToWait = 10;

    // task
    TaskHandle  taskHandle = 0;
    DAQmxCreateTask("TestNIDAQmxAO_Finite", &taskHandle);
    DAQmxCreateAOVoltageChan(taskHandle, physicalChannel, nameToAssignChannel, minVal, maxVal, units, customScaleName);
    DAQmxCfgSampClkTiming(taskHandle, source, rate, activeEdge, sampleMode, sampsPerChanToAquire);
    DAQmxWriteAnalogF64(taskHandle, sampsPerChanToAquire, autoStart, timeout, dataLayout, writeArray, &sampsPerChanWritten, NULL);
    DAQmxStartTask(taskHandle);
    DAQmxWaitUntilTaskDone(taskHandle, timeToWait);

    if ( taskHandle != 0 ) {
        DAQmxStopTask(taskHandle);
        DAQmxClearTask(taskHandle);
    }
}

int main() {
    testCode();

    int number = 1000;

    clock_t startTime = clock();
    for (int i = 0; i < number; i++) {
        testCode();
    }
    double elapsedTime = (double)(clock() - startTime) / CLOCKS_PER_SEC;
    
    printf("Test code was run %d times\n", number);
    printf("Total runetime: %f seconds\n", elapsedTime);
    printf("Average run time: %f seconds\n", elapsedTime/number);

    return 0;
}

File: MigrationVito/Python/TestNIDAQmxAO_Finite.py
import timeit

setup = '''
import nidaqmx
import numpy as np

# add_ao_voltage_chan options
physical_channel = "PXI1Slot5/ao0"
name_to_assign_to_channel = "AO Voltage"
min_val = -10
max_val = 10
units = nidaqmx.constants.VoltageUnits.VOLTS

# cfg_samp_clk_timing options
rate = 10000
source = "OnboardClock"
active_edge = nidaqmx.constants.Edge.RISING
sample_mode = nidaqmx.constants.AcquisitionType.FINITE
samps_per_chan = 100

# wave form configuration
freq = 1000
amp = 5
times = np.linspace(0, samps_per_chan / rate, samps_per_chan)

# write options
data = amp * np.sin(2*np.pi * freq * times)
auto_start = False
timeout = 10.0
'''

stmt = '''
# create task
with nidaqmx.Task('TestNIDAQmxAO_Finite') as task:
    task.ao_channels.add_ao_voltage_chan(physical_channel, name_to_assign_to_channel, min_val, max_val, units)
    task.timing.cfg_samp_clk_timing(rate, source, active_edge, sample_mode, samps_per_chan)
    task.write(data, auto_start, timeout)
    task.start()
    task.wait_until_done(timeout)
    task.stop()
'''

number = 1000

time = timeit.timeit(stmt, setup, number=number) 

print("Your code was run", number, "times")
print("Total run time:", time, "seconds")
print("Average runtime:", time/number, "seconds")